----
string[] distal_links
string[] base_links